File: components/deno/Typing.ts
import * as Sets from "./Set.ts";
import * as Maps from "./Map.ts";

export type Var = string;

export interface Type {
  apply: (s: Subst) => Type;
  ftv: () => Set<Var>;
}

export class TVar implements Type {
  name: Var;

  constructor(name: Var) {
    this.name = name;
  }

  apply(s: Subst): Type {
    return s.get(this.name) || this;
  }

  ftv(): Set<Var> {
    return new Set([this.name]);
  }

  toString(): string {
    return this.name;
  }
}

export class TCon implements Type {
  name: string;
  args: Array<Type>;

  constructor(
    name: string,
    args: Array<Type> = [],
  ) {
    this.name = name;
    this.args = args;
  }

  apply(s: Subst): Type {
    if (this.args.length === 0) {
      return this;
    }

    return new TCon(this.name, applyArray(s, this.args));
  }
  ftv(): Set<Var> {
    return Sets.flatUnion(this.args.map((t) => t.ftv()));
  }

  toString(): string {
    return `${this.name}${this.args.length > 0 ? " " : ""}${
      this.args.map((t) =>
        (t instanceof TCon && t.args.length > 0 || t instanceof TArr)
          ? `(${t.toString()})`
          : t.toString()
      ).join(" ")
    }`;
  }
}

export class TArr implements Type {
  domain: Type;
  range: Type;

  constructor(domain: Type, range: Type) {
    this.domain = domain;
    this.range = range;
  }

  apply(s: Subst): Type {
    return new TArr(this.domain.apply(s), this.range.apply(s));
  }

  ftv(): Set<Var> {
    return new Set([...this.domain.ftv(), ...this.range.ftv()]);
  }

  toString(): string {
    if (this.domain instanceof TArr) {
      return `(${this.domain}) -> ${this.range}`;
    } else {
      return `${this.domain} -> ${this.range}`;
    }
  }
}

export class TTuple implements Type {
  types: Type[];

  constructor(types: Type[]) {
    this.types = types;
  }

  apply(s: Subst): Type {
    return new TTuple(applyArray(s, this.types));
  }

  ftv(): Set<Var> {
    return new Set(this.types.flatMap((t) => [...t.ftv()]));
  }

  toString(): string {
    return `(${this.types.join(" * ")})`;
  }
}

export const typeBool = new TCon("Bool");
export const typeError = new TCon("Error");
export const typeInt = new TCon("Int");
export const typeString = new TCon("String");
export const typeUnit = new TCon("()");

export class Subst {
  protected items: Map<Var, Type>;

  constructor(items: Map<Var, Type>) {
    this.items = items;
  }

  compose(s: Subst): Subst {
    return new Subst(
      Maps.union(Maps.map(s.items, (v) => v.apply(this)), this.items),
    );
  }

  get(v: Var): Type | undefined {
    return this.items.get(v);
  }

  entries(): Array<[Var, Type]> {
    return [...this.items.entries()];
  }

  remove(names: Set<Var>): Subst {
    return new Subst(Maps.removeKeys(this.items, names));
  }
}

export const nullSubs = new Subst(new Map());

export class Scheme {
  names: Set<Var>;
  type: Type;

  constructor(names: Set<Var>, type: Type) {
    this.names = names;
    this.type = type;
  }

  apply(s: Subst): Scheme {
    return new Scheme(this.names, this.type.apply(s.remove(this.names)));
  }

  ftv(): Set<Var> {
    return Sets.difference(this.type.ftv(), new Set(this.names));
  }

  instantiate(pump: Pump): Type {
    const subst = new Subst(
      new Map([...this.names].map((n) => [n, pump.next()])),
    );

    return this.type.apply(subst);
  }

  toString(): string {
    return `${
      this.names.size == 0 ? "" : `∀ ${[...this.names].join(", ")}. `
    }${this.type}`;
  }
}

export const applyArray = (s: Subst, ts: Array<Type>): Array<Type> =>
  ts.map((t) => t.apply(s));

interface DataConstructor {
  name: string;
  args: Array<Type>;
}

export class DataDefinition {
  name: string;
  parameters: Array<string>;
  constructors: Array<DataConstructor>;

  constructor(
    name: string,
    parameters: Array<string>,
    constructors: Array<DataConstructor>,
  ) {
    this.name = name;
    this.parameters = parameters;
    this.constructors = constructors;
  }

  toString(): string {
    return `${this.name}${this.parameters.length > 0 ? " " : ""}${
      this.parameters.join(" ")
    } = ${this.constructors.map((c) => c.toString()).join(" | ")}`;
  }

  instantiate(pump: Pump): Subst {
    return new Subst(
      new Map([...this.parameters].map((n) => [n, pump.next()])),
    );
  }
}

export class TypeEnv {
  values: Map<string, Scheme>;
  adts: Array<DataDefinition>;

  constructor(values: Map<string, Scheme>, adts: Array<DataDefinition>) {
    this.values = values;
    this.adts = adts;
  }

  extend(name: string, scheme: Scheme): TypeEnv {
    const result = Maps.clone(this.values);

    result.set(name, scheme);

    return new TypeEnv(result, this.adts);
  }

  addData(adt: DataDefinition): TypeEnv {
    const adts = [adt, ...this.adts.filter((a) => a.name !== adt.name)];

    return new TypeEnv(this.values, adts);
  }

  findConstructor(
    name: string,
  ): [DataConstructor, DataDefinition] | undefined {
    for (const adt of this.adts) {
      for (const constructor of adt.constructors) {
        if (constructor.name === name) {
          return [constructor, adt];
        }
      }
    }

    return undefined;
  }

  apply(s: Subst): TypeEnv {
    return new TypeEnv(
      Maps.map(this.values, (scheme) => scheme.apply(s)),
      this.adts,
    );
  }

  ftv(): Set<Var> {
    return Sets.flatUnion([...this.values.values()].map((v) => v.ftv()));
  }

  scheme(name: string): Scheme | undefined {
    return this.values.get(name);
  }

  data(name: string): DataDefinition | undefined {
    return this.adts.find((a) => a.name === name);
  }

  generalise(t: Type): Scheme {
    return new Scheme(Sets.difference(t.ftv(), this.ftv()), t);
  }
}

export const emptyTypeEnv = new TypeEnv(new Map(), []);

export type Pump = { next: () => TVar; nextN: (n: number) => Array<TVar> };

export const createFresh = (): Pump => {
  let count = 0;

  return {
    next: (): TVar => new TVar("V" + ++count),
    nextN: (n: number): Array<TVar> =>
      Array(n).fill(0).map(() => new TVar("V" + ++count)),
  };
};

File: components/kotlin/app/src/main/kotlin/tlca/Typing.kt
package tlca

typealias Var = String

sealed class Type {
    abstract fun apply(s: Subst): Type
    abstract fun ftv(): Set<Var>
}

data class TVar(val name: String) : Type() {
    override fun apply(s: Subst): Type =
        s[name] ?: this

    override fun ftv(): Set<Var> =
        setOf(name)

    override fun toString(): String = name
}

data class TCon(val name: String, val args: List<Type> = emptyList()) : Type() {
    override fun apply(s: Subst): Type =
        if (args.isEmpty()) this else TCon(name, args.map { it.apply(s) })

    override fun ftv(): Set<Var> = args.fold(emptySet()) { acc, t -> acc + t.ftv() }

    override fun toString(): String = if (args.isEmpty()) name else "$name ${
        args.joinToString(" ") { if (it is TCon && it.args.isNotEmpty() || it is TArr) "($it)" else "$it" }
    }"
}

data class TTuple(val types: List<Type>) : Type() {
    override fun apply(s: Subst): Type =
        TTuple(types.map { it.apply(s) })

    override fun ftv(): Set<Var> =
        types.fold(emptySet()) { acc, t -> acc + t.ftv() }

    override fun toString(): String = "(${types.joinToString(" * ")})"
}

data class TArr(val domain: Type, val range: Type) : Type() {
    override fun apply(s: Subst): Type =
        TArr(domain.apply(s), range.apply(s))

    override fun ftv(): Set<Var> =
        domain.ftv() + range.ftv()

    override fun toString(): String =
        if (domain is TArr) "($domain) -> $range" else "$domain -> $range"
}

val typeBool = TCon("Bool")
val typeError = TCon("Error")
val typeInt = TCon("Int")
val typeString = TCon("String")
val typeUnit = TCon("()")

data class Subst(private val items: Map<Var, Type>) {
    infix fun compose(s: Subst): Subst =
        Subst(s.items.mapValues { it.value.apply(this) } + items)

    operator fun get(v: Var): Type? = items[v]

    operator fun minus(names: Set<Var>): Subst =
        Subst(items - names)
}

val nullSubst = Subst(emptyMap())

data class Scheme(private val names: Set<Var>, private val type: Type) {
    fun apply(s: Subst): Scheme =
        Scheme(names, type.apply(s - names))

    fun ftv(): Set<Var> =
        type.ftv() - names

    fun instantiate(pump: Pump): Type =
        type.apply(Subst(names.toList().associateWith { pump.next() }))
}

data class DataConstructor(
    val name: String,
    val args: List<Type>
) {
    val arity: Int
        get() = args.size
}

data class DataDefinition(val name: String, val typeVars: List<String>, val constructors: List<DataConstructor>) {
    override fun toString(): String = "$name${if (typeVars.isEmpty()) "" else " "}${typeVars.joinToString(" ")} = ${
        constructors.joinToString(" | ") {
            "${it.name}${if (it.args.isEmpty()) "" else " "}${
                it.args.joinToString(" ") { if (it is TCon && it.args.isNotEmpty() || it is TArr) "($it)" else "$it" }
            }"
        }
    }"
}

data class TypeEnv(private val values: Map<String, Scheme>, private val adts: List<DataDefinition>) {
    private val ftv = values.toList().flatMap { it.second.ftv() }.toSet()

    fun extend(name: String, scheme: Scheme): TypeEnv =
        TypeEnv(values + Pair(name, scheme), adts)

    fun addData(data: DataDefinition): TypeEnv =
        TypeEnv(values, adts.filter { it.name != data.name } + data)

    fun findConstructor(name: String): Pair<DataConstructor, DataDefinition>? =
        adts.flatMap { it.constructors.map { c -> Pair(c, it) } }.find { it.first.name == name }

    operator fun plus(v: Pair<String, Scheme>): TypeEnv =
        this.extend(v.first, v.second)

    operator fun plus(v: List<Pair<String, Scheme>>): TypeEnv =
        v.fold(this) { acc, p -> acc + p }

    fun apply(s: Subst): TypeEnv =
        TypeEnv(values.mapValues { it.value.apply(s) }, adts)

    operator fun get(name: String): Scheme? = values[name]

    fun data(name: String): DataDefinition? = adts.find { it.name == name }

    fun generalise(type: Type): Scheme =
        Scheme(type.ftv() - ftv, type)
}

val emptyTypeEnv = TypeEnv(emptyMap(), emptyList())
val defaultTypeEnv = emptyTypeEnv
    .extend("string_length", Scheme(setOf(), TArr(typeString, typeInt)))
    .extend("string_concat", Scheme(setOf(), TArr(typeString, TArr(typeString, typeString))))
    .extend("string_substring", Scheme(setOf(), TArr(typeString, TArr(typeInt, TArr(typeInt, typeString)))))
    .extend("string_equal", Scheme(setOf(), TArr(typeString, TArr(typeString, typeBool))))
    .extend("string_compare", Scheme(setOf(), TArr(typeString, TArr(typeString, typeInt))))


data class Pump(private var counter: Int = 0) {
    fun next(): TVar {
        counter += 1
        return TVar("V$counter")
    }

    fun nextN(size: Int): List<TVar> =
        (1..size).map { next() }
}
